src/main/host/descriptor/socket/netlink.rs

```diff
@@ -0,0 +1,1215 @@
+                    address: Ipv4Addr::LOCALHOST,
+                    label: String::from("lo"),
+                    prefix_len: 8,
+                    if_type: Arphrd::Loopback,
+                    mtu: c::CONFIG_MTU,
+                    scope: RtScope::Host,
+                    index: 1,
+                },
+                Interface {
+                    address: default_ip,
+                    label: String::from("eth0"),
+                    prefix_len: 24,
+                    if_type: Arphrd::Ether,
+                    mtu: c::CONFIG_MTU,
+                    scope: RtScope::Universe,
+                    index: 2,
+                },
+            ];
+
+            let mut common = NetlinkSocketCommon {
+                buffer,
+                send_limit: NETLINK_SOCKET_DEFAULT_BUFFER_SIZE,
+                sent_len: 0,
+                event_source: StateEventSource::new(),
+                state: FileState::ACTIVE,
+                status,
+                has_open_file: false,
+                interfaces,
+            };
+            let protocol_state = ProtocolState::new(&mut common, weak);
+            AtomicRefCell::new(Self {
+                common,
+                protocol_state,
+            })
+        })
+    }
+
+    pub fn status(&self) -> FileStatus {
+        self.common.status
+    }
+
+    pub fn set_status(&mut self, status: FileStatus) {
+        self.common.status = status;
+    }
+
+    pub fn mode(&self) -> FileMode {
+        FileMode::READ | FileMode::WRITE
+    }
+
+    pub fn has_open_file(&self) -> bool {
+        self.common.has_open_file
+    }
+
+    pub fn supports_sa_restart(&self) -> bool {
+        self.common.supports_sa_restart()
+    }
+
+    pub fn set_has_open_file(&mut self, val: bool) {
+        self.common.has_open_file = val;
+    }
+
+    pub fn getsockname(&self) -> Result<Option<nix::sys::socket::NetlinkAddr>, SyscallError> {
+        self.protocol_state.bound_address()
+    }
+
+    pub fn getpeername(&self) -> Result<Option<nix::sys::socket::NetlinkAddr>, SyscallError> {
+        log::warn!("getpeername() syscall not yet supported for netlink sockets; Returning ENOSYS");
+        Err(Errno::ENOSYS.into())
+    }
+
+    pub fn address_family(&self) -> nix::sys::socket::AddressFamily {
+        nix::sys::socket::AddressFamily::Netlink
+    }
+
+    pub fn close(&mut self, cb_queue: &mut CallbackQueue) -> Result<(), SyscallError> {
+        self.protocol_state.close(&mut self.common, cb_queue)
+    }
+
+    fn refresh_file_state(&mut self, cb_queue: &mut CallbackQueue) {
+        self.protocol_state
+            .refresh_file_state(&mut self.common, cb_queue)
+    }
+
+    pub fn shutdown(
+        &mut self,
+        _how: Shutdown,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<(), SyscallError> {
+        // We follow the same approach as UnixSocket
+        log::warn!("shutdown() syscall not yet supported for netlink sockets; Returning ENOSYS");
```

Comment:
We have been using one of the macros `warn_once_then_debug()` or `warn_one_then_trace()` for code paths where Shadow does not yet support all features that the application expects to be available in Linux, so that we don't end up spamming the log file if an application calls an unsupported function over and over but then falls back to something else that does work.

Example here:

https://github.com/shadow/shadow/blob/ab8f38e7f5b394035c6c7f3e12a7d66df100ac21/src/main/host/syscall/handler/fcntl.rs#L175-L178

Notice that we tend to prefix these messages with `(LOG_ONCE) ` to make it clear in the logs that this particular message is not going to be repeated.

Note that `log::warn!()` should still be used in places where an application is legitimately sending an invalid arg or something.

I noticed several places throughout your PR where the warn once pattern should be applied, but I'm only going to comment here once to avoid the extra spam. Please scan through your usage of `log::warn!()` to see which should be changed to the warn once pattern.

Reply:
The unix socket is my main reference for implementing the netlink socket. I see what I did a lot in the unix socket https://github.com/shadow/shadow/blob/97d99/src/main/host/descriptor/socket/unix.rs

Is it supposed to be changed later in the unix socket as well?

Reply:
Yes, we would ideally update other usages inside the Shadow codebase too. (We only just recently added the "log once" functionality in 82e59bc but we didn't thoroughly go through and fix all log lines yet.)